1- Launch the CoherenceTestIwor.py program with default dataset used in the Python code that is TokenVieuxM.txt 

* a) Launch the program several times. What did you observe regarding the obtained coherence and perplexity values ? Try to provide a reason for your observation (looking to papers or Web infos referring to LDA topic identification process). Consequently, explain how it can be possible to obtain the best results regarding coherence and perplexity.

* b) Look the the topic descriptions obtained with one specific launch. Present the topic description output you obtain along with coherence and perplexity values. What can you conclude regarding these descriptions and theses values ? Explain if you consider that the results are good are bad. Justify your answer in different ways.

In [1]:
# Import dependecies
import nltk
# Create a corpus from a list of lists of tokens
from gensim.corpora.dictionary import Dictionary
from nltk.corpus import stopwords

nltk.download('stopwords')


[nltk_data] Downloading package stopwords to /home/dario/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

In [2]:
texts=[]
file = open("TokenVieuxM.txt", "r")
lines = file.readlines()
file.close()


In [3]:
# Clean the data
for line in lines:
    line=line.strip()
    lt=line.split(",")
    #Potential ill-character cleaning
    for i in range(len(lt)):
        lt[i]=lt[i].replace('[','')
        lt[i]=lt[i].replace(']','')
        lt[i]=lt[i].replace('"','')
        lt[i]=lt[i].replace('\n','')
        lt[i]=lt[i].replace(' ', '')
    #End : Potential ill-characters cleaning
    # print(lt)
    # ltc=[word for word in lt if not word in stopwords.words()]
    # print("C", ltc)
    texts.append(lt)


In [18]:
#Here set the number of topics(to be changed if necessary)
nb=10
  
id2word = Dictionary(texts)
corpus = [id2word.doc2bow(text) for text in texts]
#print(corpus)

# # Print dictionnary
# for i in id2word:
#   print(i, id2word[i])


In [19]:
# Train the lda model on the corpus.
from gensim.models import LdaModel

lda = LdaModel(corpus, num_topics=nb)


In [20]:
# Print topic descrition
for i in range(0, nb):
  value=lda.get_topic_terms(i)
#  print(value)
  print("Topic ", i)
  for j in value:
    print(id2word[j[0]], " - P=", j[1])
  print()

Topic  0
of  - P= 0.032965027
and  - P= 0.032303113
to  - P= 0.030246342
the  - P= 0.029939067
in  - P= 0.021617722
with  - P= 0.014988868
their  - P= 0.014757316
structural  - P= 0.013033258
these  - P= 0.0125924
strategies  - P= 0.011675524

Topic  1
the  - P= 0.029451385
and  - P= 0.026933555
of  - P= 0.019172484
auditory  - P= 0.017929075
in  - P= 0.012719797
neurons  - P= 0.010169702
to  - P= 0.009143855
are  - P= 0.008719574
various  - P= 0.0070741978
audiomotor  - P= 0.0058763027

Topic  2
and  - P= 0.012832806
the  - P= 0.010964359
of  - P= 0.010058744
to  - P= 0.005431109
in  - P= 0.0052027926
auditory  - P= 0.0051140897
neurons  - P= 0.004636731
are  - P= 0.0046162033
biological  - P= 0.00418508
for  - P= 0.0039869635

Topic  3
of  - P= 0.053148005
the  - P= 0.049691465
and  - P= 0.039368074
to  - P= 0.026892057
in  - P= 0.023397615
auditory  - P= 0.010968063
a  - P= 0.010241951
for  - P= 0.008651338
are  - P= 0.008533108
neurons  - P= 0.007932067

Topic  4
the  - P= 0.033451

In [21]:
# Compute Perplexity
perplexity_lda=lda.log_perplexity(corpus)  # a measure of how good the model is (lower the better).
print('Perplexity= ', perplexity_lda)


Perplexity=  -7.526049896583079


In [22]:
# Compute Coherence Score
from gensim.models import CoherenceModel

coherence_model_lda = CoherenceModel(model=lda, texts=texts, dictionary=id2word, coherence='c_v')
coherence_lda = coherence_model_lda.get_coherence()
print('Coherence= ', coherence_lda)

Coherence=  0.33459597719427553


In [23]:
def train_model_and_compute_metrics(nb):

    # Train the lda model on the corpus.
    from gensim.models import LdaModel

    lda = LdaModel(corpus, num_topics=nb)

    # Print topic descrition
    # for i in range(0, nb-1):
    #     value=lda.get_topic_terms(i)
    #     #  print(value)
    #     print("Topic ", i)
    #     for j in value:
    #         print(id2word[j[0]], " - P=", j[1])
    #     print()

    # Compute Perplexity
    perplexity_lda=lda.log_perplexity(corpus)  # a measure of how good the model is (lower the better).
    # print('Perplexity= ', perplexity_lda)

    # Compute Coherence Score
    from gensim.models import CoherenceModel

    coherence_model_lda = CoherenceModel(model=lda, texts=texts, dictionary=id2word, coherence='c_v')
    coherence_lda = coherence_model_lda.get_coherence()
    # print('Coherence= ', coherence_lda)
    return perplexity_lda, coherence_lda



In [24]:
perplexities = []
coherences = []
for i in range(50):
    p, c = train_model_and_compute_metrics(10)
    perplexities.append(p)
    coherences.append(c)

from statistics import mean

print(perplexities)
print(coherences)
print(f"Min perplexity: {min(perplexities)}")
print(f"Max perplexity: {max(perplexities)}")
print(f"Average Perplexity: {mean(perplexities)}")
print(f"Min coherence: {min(coherences)}")
print(f"Max coherence: {max(coherences)}")
print(f"Average Coherence:  {mean(coherences)}")

[-7.191772159795034, -7.522204059721802, -7.437199070320737, -7.314576423751345, -7.199724123000556, -7.442886027595424, -7.353710299262414, -7.48433370455947, -6.941227889237279, -7.187834788236738, -7.438033253734294, -7.52544761131241, -7.544524989233896, -7.009913421938443, -7.702930060534212, -7.032278449190897, -7.140420978115416, -7.270083453332469, -7.471877117409234, -7.638414137599411, -7.189738856105132, -7.025425739255781, -7.3365646965538955, -7.414219959468429, -7.410984829627613, -7.232060893675031, -7.179793820378452, -7.2741931242205045, -7.0243961364200365, -7.544641575240981, -7.629902277044333, -7.111960665832472, -7.409063697268125, -7.318910627031489, -7.358435815172662, -7.313154924783826, -7.466702046044451, -7.169366885370769, -7.411112549972208, -7.408851414545946, -7.436131601786586, -7.559518204207307, -7.306268895716662, -6.986136560282745, -7.303805437104287, -7.511955677310219, -7.460578292574357, -7.61724541184574, -7.156611029944457, -7.189920338347504]

# 1-a)

We run the program 50 times, here are some observations:

Perplexity:
- The minimum perplexity score obtained was -7.702930060534212, while the maximum was -6.941227889237279. The range between the minimum and maximum values suggests that there is some variation in the model's ability to predict the held-out data.
- The average perplexity score across all runs was -7.33214088002027. This average value can serve as a baseline for comparing the perplexity of future runs or different LDA configurations.

Coherence:
- The coherence scores range from 0.293095324769671 to 0.4802758872561606, indicating a noticeable variation in the interpretability and semantic consistency of the generated topics across the different runs.
- The average coherence score across all runs was 0.3896210491840405. This average value can serve as a baseline for comparing the coherence of future runs or different LDA configurations.

Overall, it appears that there is some variability in the performance of the LDA model across different runs. Some runs yield higher coherence scores, indicating more interpretable and coherent topics, while others have lower coherence scores. Similarly, there is some variation in the perplexity scores, suggesting differences in the model's ability to predict the held-out data.

The observed variation in coherence and perplexity scores can be attributed to several factors related to the LDA topic identification process. Here are some reasons for the observed observations:


1. Topic Overlap and Ambiguity: In the field of healthcare research, where the corpus consists of abstracts from research papers, there might be inherent topic overlap and ambiguity. Some topics may share common terms or concepts, making it challenging for the model to accurately distinguish between them. This can impact the coherence and perplexity scores, as the model struggles to assign tokens to distinct topics.

2. Variability in Data Quality: The quality and diversity of the research papers' abstracts can also influence the LDA model's performance. If the dataset contains papers with varying levels of quality, inconsistency, or noise, it can affect the coherence and perplexity scores. The presence of irrelevant or poorly written abstracts can introduce noise and decrease the model's ability to identify coherent topics.

3. Random Initialization: LDA models use random initialization, which means that each run can start from a different initial state. Randomness can lead to different outcomes for each run, resulting in variations in the coherence and perplexity scores. It's possible that some runs initialized with favorable initial conditions led to better topic quality and lower perplexity, while others had less favorable initial states.


To obtain the best results in terms of coherence and perplexity for an LDA model, we can consider the following strategies:

1. Hyperparameter Tuning: Experiment with different hyperparameter configurations to find the optimal settings for the dataset. This includes adjusting the number of topics, alpha, and beta values. We can try running the model with various combinations of these parameters and select the configuration that yields the highest coherence and lowest perplexity scores. Techniques like grid search or Bayesian optimization can be helpful in automating this process.

2. Quality Preprocessing: Ensuring that data preprocessing steps are robust. This includes removing irrelevant or noisy data, such as stopwords, punctuation, and rare terms. Applying stemming or lemmatization techniques can also help reduce the dimensionality of the corpus and improve the quality of the topics generated by the model.

3. Text Cleaning: Perform thorough text cleaning to eliminate any inconsistencies, formatting issues, or special characters that might affect the coherence and perplexity scores. Standardizing the text, handling missing values, and resolving any encoding problems are crucial steps to ensure the input data is clean and consistent.

4. Dataset Size and Diversity: Consider the size and diversity of the dataset. Larger and more diverse datasets tend to yield better results. If the dataset is small, may be wanted to augment it or consider techniques like cross-validation to assess the stability and generalizability of the generated topics.

5. Iterations and Convergence: Increasing the number of iterations during the training process. More iterations allow the model to converge to a more stable state and potentially improve the coherence and perplexity scores. However, it should be take in account the computational resources required for longer training times.

6. Evaluation Metrics: While coherence and perplexity are commonly used evaluation metrics, they may not always align with human judgment. It's important to complement these metrics with manual inspection and validation of the generated topics. Human evaluation can provide valuable insights into the interpretability and relevance of the topics.

7. Ensembling or Averaging: Instead of relying on a single run, is possible to perform ensembling or averaging of multiple LDA models. This involves running the LDA model multiple times with different random initializations and combining the results to get a more robust representation of the topics. Ensembling or averaging can help mitigate the variability observed in coherence and perplexity scores across individual runs.

The topic modeling is an iterative process, and it may require several iterations and experimentation to achieve the best results. It's important to strike a balance between model complexity, computational resources, and the interpretability of the generated topics based on the specific requirements of the domain.


In [28]:
nb = 10

# Train the lda model on the corpus.
from gensim.models import LdaModel

lda = LdaModel(corpus, num_topics=nb)

# Print topic descrition
for i in range(0, nb):
  value=lda.get_topic_terms(i)
#  print(value)
  print("Topic ", i)
  for j in value:
    print(id2word[j[0]], " - P=", j[1])
  print()

# Compute Perplexity
perplexity_lda=lda.log_perplexity(corpus)  # a measure of how good the model is (lower the better).
print('Perplexity= ', perplexity_lda)

# Compute Coherence Score
from gensim.models import CoherenceModel

coherence_model_lda = CoherenceModel(model=lda, texts=texts, dictionary=id2word, coherence='c_v')
coherence_lda = coherence_model_lda.get_coherence()
print('Coherence= ', coherence_lda)

Topic  0
and  - P= 0.04079648
the  - P= 0.03669876
of  - P= 0.035698477
to  - P= 0.020153081
with  - P= 0.015370513
in  - P= 0.014031986
their  - P= 0.011863557
are  - P= 0.008931493
for  - P= 0.008719349
strategies  - P= 0.0074828193

Topic  1
the  - P= 0.042128753
and  - P= 0.037088867
of  - P= 0.036617726
to  - P= 0.020170096
in  - P= 0.01629544
auditory  - P= 0.015384982
neurons  - P= 0.011424743
are  - P= 0.0081757605
for  - P= 0.0075630527
with  - P= 0.007406899

Topic  2
the  - P= 0.041734356
of  - P= 0.041273266
and  - P= 0.03333703
in  - P= 0.028039215
to  - P= 0.027878301
a  - P= 0.011835274
that  - P= 0.0095071765
for  - P= 0.009258195
we  - P= 0.00742808
are  - P= 0.0073413337

Topic  3
the  - P= 0.032281876
of  - P= 0.027554318
a  - P= 0.027267015
to  - P= 0.02571993
in  - P= 0.017915018
is  - P= 0.017450055
for  - P= 0.014408661
and  - P= 0.012528666
protein  - P= 0.011475687
proteomics  - P= 0.010590326

Topic  4
and  - P= 0.044136
of  - P= 0.039803825
the  - P= 0.036199

# 1- b)

```
Topic  0
of  - P= 0.011058775
the  - P= 0.009887789
in  - P= 0.009401254
to  - P= 0.006715426
and  - P= 0.0062425523
ice  - P= 0.0060687326
that  - P= 0.005530666
evidence  - P= 0.0053881146
months  - P= 0.005122016
age  - P= 0.004753692

Topic  1
and  - P= 0.003016283
the  - P= 0.0027025968
of  - P= 0.0026951975
neurons  - P= 0.0024320884
auditory  - P= 0.0024301207
to  - P= 0.0024213113
in  - P= 0.0023258585
for  - P= 0.0023029167
are  - P= 0.0022783382
has  - P= 0.0022629541

Topic  2
of  - P= 0.04065094
and  - P= 0.03873536
the  - P= 0.03197248
to  - P= 0.025909802
in  - P= 0.016355481
a  - P= 0.015465702
for  - P= 0.015459485
on  - P= 0.010814473
is  - P= 0.008926802
biological  - P= 0.0070602344

Topic  3
of  - P= 0.04618509
the  - P= 0.04207941
and  - P= 0.03924803
to  - P= 0.02932435
in  - P= 0.020202585
with  - P= 0.016553894
these  - P= 0.013052856
their  - P= 0.012730474
strategies  - P= 0.011087511
men  - P= 0.011045343

Topic  4
the  - P= 0.04434207
of  - P= 0.036716737
and  - P= 0.024621978
to  - P= 0.024028737
a  - P= 0.023211565
in  - P= 0.018962674
is  - P= 0.013509136
for  - P= 0.010867239
protein  - P= 0.00973452
dna  - P= 0.009628494

Topic  5
of  - P= 0.032803472
the  - P= 0.02763419
and  - P= 0.023926117
in  - P= 0.022081278
to  - P= 0.01887759
for  - P= 0.007857354
a  - P= 0.007245258
that  - P= 0.0072033205
months  - P= 0.0069121956
are  - P= 0.0063646236

Topic  6
and  - P= 0.0462492
the  - P= 0.039982613
of  - P= 0.030670736
in  - P= 0.024383793
to  - P= 0.019571112
neurons  - P= 0.013130104
auditory  - P= 0.0111098355
for  - P= 0.007851063
are  - P= 0.0076545
that  - P= 0.007559434

Topic  7
the  - P= 0.05403955
of  - P= 0.026726577
and  - P= 0.025404995
auditory  - P= 0.023493377
to  - P= 0.017092828
in  - P= 0.013509813
neurons  - P= 0.011250108
are  - P= 0.010783905
for  - P= 0.0077176117
be  - P= 0.007553489

Topic  8
of  - P= 0.03431679
the  - P= 0.032735877
and  - P= 0.027877424
in  - P= 0.0225479
to  - P= 0.01929094
that  - P= 0.009479902
ice  - P= 0.009157433
a  - P= 0.008141336
dna  - P= 0.006896983
neurons  - P= 0.006812114

Topic  9
of  - P= 0.022648524
and  - P= 0.018146789
the  - P= 0.017388334
to  - P= 0.01587528
with  - P= 0.009157343
their  - P= 0.00729402
for  - P= 0.007259891
are  - P= 0.0066940407
structural  - P= 0.0066021215
in  - P= 0.0059986953

Perplexity=  -7.244668698758395
Coherence=  0.39542486916238984
```

Based on the topics and the corresponding perplexity and coherence values, we can analyze the results as follows:

1. Topic Descriptions: The topic descriptions consist of the top words and their corresponding probabilities for each topic. From a qualitative perspective, it is difficult to determine the specific theme or subject of each topic based solely on the top words. The words appear to be common and generic, such as "of," "the," "and," "to," etc. However, there are some topic-specific terms like "ice," "neurons," "auditory," "protein," "DNA," etc., which suggest potential themes related to neuroscience, genetics, and biological research.

2. Perplexity: The perplexity value indicates how well the LDA model predicts the held-out test data. A lower perplexity score generally suggests better model performance. In this case, the perplexity value of -7.2446 is lower than the average perplexity (-7.3321) obtained from the 50 runs. This indicates that the model performs relatively well in terms of predicting unseen data, suggesting that the model has learned the underlying patterns of the corpus to some extent.

3. Coherence: The coherence value measures the degree of semantic similarity among the top words within each topic. A higher coherence score indicates that the topics are more coherent and interpretable. The coherence score of 0.3954 is higher than the average coherence (0.3896) obtained from the 50 runs. This suggests that the generated topics have a reasonable degree of semantic coherence.

Considering these observations, we can conclude the following:

a) Results Assessment: Based on the perplexity and coherence values, the obtained results can be considered relatively good. The perplexity score is lower than the average, indicating that the model performs well in predicting unseen data. The coherence score is higher than the average, suggesting that the topics exhibit a reasonable level of semantic coherence.

b) Interpretability: However, when examining the topic descriptions, the top words for each topic appear to be quite generic, including common words like "of," "the," and "and," which do not provide clear thematic information. This indicates that the model may not have captured specific and distinct topics within the healthcare research domain. Removing stopwords could highly improve the results.

c) Domain Relevance: It is important to assess the relevance and interpretability of the topics within the specific context of healthcare research. The generic nature of the top words raises questions about the specificity and meaningfulness of the generated topics. Further evaluation and validation from domain experts are necessary to determine the relevance and usefulness of the identified topics for healthcare research.

d) Comparison to Other Runs: To make a more comprehensive judgment, it would be valuable to compare these results with the average perplexity and coherence values obtained from the full set of 50 runs. Additionally, visualizing the topics using techniques like word clouds or examining topic-document distributions could provide additional insights into the quality and meaningfulness of the topics.

In summary, while the perplexity and coherence scores suggest relatively good model performance, the interpretability and specificity of the generated topics need further evaluation. Additional iterations, hyperparameter tuning, or incorporating domain-specific knowledge may be necessary to improve the results and generate more meaningful and relevant topics for healthcare research.